server/src/main/java/io/crate/fdw/JdbcForeignDataWrapper.java

```diff
@@ -142,14 +172,32 @@ public CompletableFuture<BatchIterator<Row>> getIterator(Role currentUser,
         RelationName remoteName = new RelationName(
             remoteSchema.isEmpty() ? foreignTable.name().schema() : remoteSchema,
             remoteTable.isEmpty() ? foreignTable.name().name() : remoteTable);
-        BatchIterator<Row> it = new JdbcBatchIterator(url, properties, refs, remoteName);
+
+        Predicate<Symbol> containsUnsafeFn = x ->
+            x instanceof Function fn && !SAFE_FUNCTIONS.contains(fn.name());
+        Predicate<Row> rowFilter = null;
+        if (SymbolVisitors.any(containsUnsafeFn, query)) {
+            // For cases like `SELECT x FROM ... WHERE y` `y` could've been pruned from refs
+            // re-add it
+            RefVisitor.visitRefs(query, ref -> {
+                if (!refs.contains(ref)) {
+                    refs.add(ref);
+                }
+            });
+            rowFilter = RowFilter.create(txnCtx, inputFactory, InputColumns.create(query, refs));
+            query = Literal.BOOLEAN_TRUE;
+        }
+        BatchIterator<Row> it = new JdbcBatchIterator(url, properties, refs, query, remoteName);
```

Comment:
It could also be an option to add a `supportsQueryPushdown(Symbol query)` function on the foreign data wrapper interface and use it in the new `MergeFilterAndForeignCollect` rule to decide if a filter should be pushed down. 

It would have the advantages that:

- push down is visible in the explain output
- fdw.getIterator doesn't have to re-implement creating a RowFilter

The downside is that there's more implicit loose coupling because the rule and the getIterator implementation. (But maybe an extra assert in the getIterator is enough to protect against mistakes there)

Reply:
Changed it to have the check if a query is supported in the rule. Advantage is that the explain output shows if the push down happens, the main downside is that it would make runtime checks on the remote server to see what scalars a server supports more difficult. (It would require IO during planning)

But we can still revisit this if this if we ever decide to query the remote system for supported scalars.